infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -175,231 +186,272 @@ public void handleTailRequest(CorfuPayloadMsg<TailsRequest> msg, ChannelHandlerC
     }
-    r.sendResponse(ctx, msg, CorfuMsgType.READ_RESPONSE.payloadMsg(rr));
-  }
-
-  @ServerHandler(type = CorfuMsgType.INSPECT_ADDRESSES_REQUEST)
-  public void inspectAddresses(
-      CorfuPayloadMsg<InspectAddressesRequest> msg, ChannelHandlerContext ctx, IServerRouter r) {
-    List<Long> addresses = msg.getPayload().getAddresses();
-    log.trace("inspectAddresses: {}", addresses);
-    InspectAddressesResponse inspectResponse = new InspectAddressesResponse();
-
-    for (long address : addresses) {
-      try {
-        if (!streamLog.contains(address)) {
-          inspectResponse.add(address);
+
+    @RequestHandler(type = PayloadCase.INSPECT_ADDRESSES_REQUEST)
+    public void handleInspectAddressesRequest(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
+        final List<Long> addresses = req.getPayload().getInspectAddressesRequest().getAddressList();
+        List<Long> emptyAddresses = new ArrayList<>();
+
+        if (log.isTraceEnabled()) {
+            log.trace("handleInspectAddressesRequest[{}]: " +
+                    "addresses {}", req.getHeader().getRequestId(), addresses);
+        }
+
+        for (long address : addresses) {
+            try {
+                if (!streamLog.contains(address)) {
+                    emptyAddresses.add(address);
+                }
+            } catch (TrimmedException te) {
+                r.sendResponse(getResponseMsg(req.getHeader(), getTrimmedErrorMsg()), ctx);
+                return;
+            } catch (DataCorruptionException dce) {
+                r.sendResponse(getResponseMsg(req.getHeader(), getDataCorruptionErrorMsg(address)), ctx);
+                return;
+            }
         }
-      } catch (TrimmedException te) {
-        r.sendResponse(ctx, msg, CorfuMsgType.ERROR_TRIMMED.msg());
-        return;
-      } catch (DataCorruptionException dce) {
-        r.sendResponse(ctx, msg, CorfuMsgType.ERROR_DATA_CORRUPTION.payloadMsg(address));
-        return;
-      }
+
+        // Note: we reuse the request header as the ignore_cluster_id and
+        // ignore_epoch fields are the same in both cases.
+        r.sendResponse(getResponseMsg(req.getHeader(), getInspectAddressesResponseMsg(emptyAddresses)), ctx);
     }
-    r.sendResponse(ctx, msg, CorfuMsgType.INSPECT_ADDRESSES_RESPONSE.payloadMsg(inspectResponse));
-  }
 
     /**
      * Handles requests for known entries in specified range.
      * This is used by state transfer to catch up only the remainder of the segment.
      */
-    @ServerHandler(type = CorfuMsgType.KNOWN_ADDRESS_REQUEST)
-    private void getKnownAddressesInRange(CorfuPayloadMsg<KnownAddressRequest> msg,
-                                          ChannelHandlerContext ctx, IServerRouter r) {
-
-        KnownAddressRequest request = msg.getPayload();
+    @RequestHandler(type = PayloadCase.KNOWN_ADDRESS_REQUEST)
+    private void handleKnownAddressRequest(RequestMsg req, ChannelHandlerContext ctx, IServerRouter r) {
```

Comment:
This handler and handleInspectAddressesRequest seem to be redundant. We should remove one of them.